// reflect-annotations/lib/cjs/annotations.js
"use strict";
///<reference types="reflect-metadata"/>
Object.defineProperty(exports, "__esModule", { value: true });
exports.createAnnotationFactory = exports.getReturnType = exports.getParameterTypes = exports.setParameterAnnotations = exports.getParameterAnnotations = exports.setAnnotations = exports.getAnnotations = exports.isAnnotationFactory = exports.AnnotationFactory = void 0;
const ANNOTATIONS = 'annotations', PARAMETER_ANNOTATIONS = 'parameter-annotations', PARAMETER_TYPES = 'design:paramtypes', RETURN_TYPE = 'design:returntype';
exports.AnnotationFactory = Symbol.for('reflect-annotations.factory');
function isAnnotationFactory(thing) {
    return typeof thing === 'function' && thing[exports.AnnotationFactory] === true;
}
exports.isAnnotationFactory = isAnnotationFactory;
function getAnnotations(target, key) {
    return Reflect.getMetadata(ANNOTATIONS, target, key) || [];
}
exports.getAnnotations = getAnnotations;
function setAnnotations(target, key, annotations) {
    Reflect.defineMetadata(ANNOTATIONS, annotations || [], target, key);
}
exports.setAnnotations = setAnnotations;
function getParameterAnnotations(target, key) {
    const annotations = Reflect.getMetadata(PARAMETER_ANNOTATIONS, target, key);
    return annotations ? Array.from(annotations) : [];
}
exports.getParameterAnnotations = getParameterAnnotations;
function setParameterAnnotations(target, key, annotations) {
    Reflect.defineMetadata(PARAMETER_ANNOTATIONS, annotations, target, key);
}
exports.setParameterAnnotations = setParameterAnnotations;
function getParameterTypes(target, key) {
    return Reflect.getMetadata(PARAMETER_TYPES, target, key);
}
exports.getParameterTypes = getParameterTypes;
function getReturnType(target, key) {
    return Reflect.getMetadata(RETURN_TYPE, target, key);
}
exports.getReturnType = getReturnType;
function createAnnotationFactory(Type) {
    return Object.assign(function annotationFactory(...args) {
        const annotationInstance = new Type(...args), annotation = (target, key, descriptorOrParamIndex) => {
            if (key && typeof descriptorOrParamIndex === 'number') {
                const annotations = getParameterAnnotations(target, key);
                annotations[descriptorOrParamIndex] = annotationInstance;
                setParameterAnnotations(target, key, annotations);
            }
            else {
                const annotations = getAnnotations(target, key);
                annotations.push(annotationInstance);
                setAnnotations(target, key, annotations);
            }
        };
        annotation.annotationInstance = annotationInstance;
        return annotation;
    }, { [exports.AnnotationFactory]: true });
}
exports.createAnnotationFactory = createAnnotationFactory;
