// reflect-annotations/lib/cjs/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AnnotationFactory = exports.isAnnotationFactory = exports.createAnnotationFactory = exports.setParameterAnnotations = exports.getParameterAnnotations = exports.setAnnotations = exports.getAnnotations = exports.reflectAnnotations = void 0;
const annotations_js_1 = require("./annotations.js");
Object.defineProperty(exports, "getAnnotations", { enumerable: true, get: function () { return annotations_js_1.getAnnotations; } });
Object.defineProperty(exports, "setAnnotations", { enumerable: true, get: function () { return annotations_js_1.setAnnotations; } });
Object.defineProperty(exports, "getParameterAnnotations", { enumerable: true, get: function () { return annotations_js_1.getParameterAnnotations; } });
Object.defineProperty(exports, "setParameterAnnotations", { enumerable: true, get: function () { return annotations_js_1.setParameterAnnotations; } });
Object.defineProperty(exports, "createAnnotationFactory", { enumerable: true, get: function () { return annotations_js_1.createAnnotationFactory; } });
Object.defineProperty(exports, "isAnnotationFactory", { enumerable: true, get: function () { return annotations_js_1.isAnnotationFactory; } });
Object.defineProperty(exports, "AnnotationFactory", { enumerable: true, get: function () { return annotations_js_1.AnnotationFactory; } });
const reflect_class_js_1 = require("./reflect-class.js");
class PropertyDescription {
    constructor(name, declaredOrder) {
        this.name = name;
        this.declaredOrder = declaredOrder;
        this.classAnnotations = [];
        this.methodAnnotations = [];
        this.parameterAnnotations = [];
        this.types = {};
    }
}
function addClassAnnotation(property, annotation) {
    property.classAnnotations.push(annotation);
    return property;
}
function addMethodAnnotation(property, annotation) {
    property.methodAnnotations.push(annotation);
    return property;
}
function collectPropertyAnnotations(property, ctor) {
    const methodAnnotations = property.name in ctor.prototype ? annotations_js_1.getAnnotations(ctor.prototype, property.name) : [], order = property.declaredOrder ? 'reduceRight' : 'reduce';
    property = annotations_js_1.getAnnotations(ctor)[order](addClassAnnotation, methodAnnotations[order](addMethodAnnotation, property));
    property.parameterAnnotations =
        property.name in ctor.prototype ? annotations_js_1.getParameterAnnotations(ctor.prototype, property.name) : [];
    property.types.parameters = annotations_js_1.getParameterTypes(ctor.prototype, property.name);
    property.types.return = annotations_js_1.getReturnType(ctor.prototype, property.name);
    return property;
}
function reflectAnnotations(source, options = { declaredOrder: true }) {
    const classMetadata = reflect_class_js_1.reflectClassProperties(source);
    return classMetadata.properties.reduce((properties, propertyName) => {
        properties.push(classMetadata.constructors.reduceRight(collectPropertyAnnotations, new PropertyDescription(propertyName, options.declaredOrder)));
        return properties;
    }, []);
}
exports.reflectAnnotations = reflectAnnotations;
exports.default = reflectAnnotations;
