// reflect-annotations/lib/esm/annotations.js
///<reference types="reflect-metadata"/>
const ANNOTATIONS = 'annotations', PARAMETER_ANNOTATIONS = 'parameter-annotations', PARAMETER_TYPES = 'design:paramtypes', RETURN_TYPE = 'design:returntype';
export const AnnotationFactory = Symbol.for('reflect-annotations.factory');
export function isAnnotationFactory(thing) {
    return typeof thing === 'function' && thing[AnnotationFactory] === true;
}
export function getAnnotations(target, key) {
    return Reflect.getMetadata(ANNOTATIONS, target, key) || [];
}
export function setAnnotations(target, key, annotations) {
    Reflect.defineMetadata(ANNOTATIONS, annotations || [], target, key);
}
export function getParameterAnnotations(target, key) {
    const annotations = Reflect.getMetadata(PARAMETER_ANNOTATIONS, target, key);
    return annotations ? Array.from(annotations) : [];
}
export function setParameterAnnotations(target, key, annotations) {
    Reflect.defineMetadata(PARAMETER_ANNOTATIONS, annotations, target, key);
}
export function getParameterTypes(target, key) {
    return Reflect.getMetadata(PARAMETER_TYPES, target, key);
}
export function getReturnType(target, key) {
    return Reflect.getMetadata(RETURN_TYPE, target, key);
}
export function createAnnotationFactory(Type) {
    return Object.assign(function annotationFactory(...args) {
        const annotationInstance = new Type(...args), annotation = (target, key, descriptorOrParamIndex) => {
            if (key && typeof descriptorOrParamIndex === 'number') {
                const annotations = getParameterAnnotations(target, key);
                annotations[descriptorOrParamIndex] = annotationInstance;
                setParameterAnnotations(target, key, annotations);
            }
            else {
                const annotations = getAnnotations(target, key);
                annotations.push(annotationInstance);
                setAnnotations(target, key, annotations);
            }
        };
        annotation.annotationInstance = annotationInstance;
        return annotation;
    }, { [AnnotationFactory]: true });
}
