// reflect-annotations/lib/esm/index.js
import { getAnnotations, setAnnotations, getParameterAnnotations, setParameterAnnotations, getParameterTypes, getReturnType, createAnnotationFactory, isAnnotationFactory, AnnotationFactory, } from './annotations.js';
import { reflectClassProperties } from './reflect-class.js';
class PropertyDescription {
    constructor(name, declaredOrder) {
        this.name = name;
        this.declaredOrder = declaredOrder;
        this.classAnnotations = [];
        this.methodAnnotations = [];
        this.parameterAnnotations = [];
        this.types = {};
    }
}
function addClassAnnotation(property, annotation) {
    property.classAnnotations.push(annotation);
    return property;
}
function addMethodAnnotation(property, annotation) {
    property.methodAnnotations.push(annotation);
    return property;
}
function collectPropertyAnnotations(property, ctor) {
    const methodAnnotations = property.name in ctor.prototype ? getAnnotations(ctor.prototype, property.name) : [], order = property.declaredOrder ? 'reduceRight' : 'reduce';
    property = getAnnotations(ctor)[order](addClassAnnotation, methodAnnotations[order](addMethodAnnotation, property));
    property.parameterAnnotations =
        property.name in ctor.prototype ? getParameterAnnotations(ctor.prototype, property.name) : [];
    property.types.parameters = getParameterTypes(ctor.prototype, property.name);
    property.types.return = getReturnType(ctor.prototype, property.name);
    return property;
}
function reflectAnnotations(source, options = { declaredOrder: true }) {
    const classMetadata = reflectClassProperties(source);
    return classMetadata.properties.reduce((properties, propertyName) => {
        properties.push(classMetadata.constructors.reduceRight(collectPropertyAnnotations, new PropertyDescription(propertyName, options.declaredOrder)));
        return properties;
    }, []);
}
export default reflectAnnotations;
export { reflectAnnotations };
export { getAnnotations, setAnnotations, getParameterAnnotations, setParameterAnnotations, createAnnotationFactory, isAnnotationFactory, AnnotationFactory, };
